package NKikimrBlobStorage;
option java_package = "ru.yandex.kikimr.proto";

// sets the parameters of PDisk
message TPDiskConfig {
    enum ESwitch {
        DoNotTouch = 0;

        Enable = 1;
        ForceEnable = 2;

        Disable = 3;
        ForceDisable = 4;
    };

    ///////////////////          Depracated           //////////////////////////
    //optional uint64 RootSchedulerQuant = 101;
    //optional uint64 OwnerSchedulerQuant = 102;
    //optional uint64 OwnerSchedulerBurst = 103;
    //optional uint64 BytesSchedulerQuant = 104;
    //optional uint64 BytesSchedulerBurst = 106;
    //optional uint64 SyncLogBurst = 115;
    //optional uint64 HugeBurst = 116;
    //optional uint64 FastReadBurst = 117;
    //optional uint64 OtherReadBurst = 118;
    //optional uint64 LoadBurst = 119;
    //optional uint64 LowReadBurst = 121;
    //optional uint64 TrimCostNs = 5; // DEPRECATED, has no effect! (Use zero to disable TRIM)
    //optional uint64 StaticGroupChunkReservePerMile = 1003;  // DEPRECATED, has no effect!
    //optional uint64 AntiLockBrakingSystemPerMile = 1002;  // 0 to disable, 1000 for full strength
    //optional uint64 HistorySize = 1004;  // History record count, displayed in http monitoring
    //optional uint64 DriveModelStepSizeBytes = 1007; // Not used anymore
    //optional uint64 UnusedBurstInflow = 1010;
    //optional uint64 LogQDurationMs = 1012;
    //optional uint64 RealtimeReadQDurationMs = 1013;
    //optional uint64 FastReadQDurationMs = 1014;
    //optional uint64 BatchQDurationMs = 1015;
    //optional uint64 TrimQDurationMs = 1016;
    //optional uint64 SensitiveQsDurationNs = 1017;
    //optional uint64 BestEffortQsDurationNs = 1018;
    //optional uint64 AdhesionSizeNs = 1019;
    //optional uint64 AdhesionLookupCost = 1020;
    //optional uint64 AdhesionLookupCount = 1021;
    ////////////////////////////////////////////////////////////////////////////

    optional uint64 StatisticsUpdateIntervalMs = 100;

    optional uint64 BytesSchedulerWeight = 105;
    optional uint64 LogWeight = 107;
    optional uint64 FreshWeight = 108;
    optional uint64 CompWeight = 109;
    optional uint64 SyncLogWeight = 110;
    optional uint64 HugeWeight = 111;
    optional uint64 FastReadWeight = 112;
    optional uint64 OtherReadWeight = 113;
    optional uint64 LoadWeight = 114;
    optional uint64 LowReadWeight = 120;

    // There will be approximately (OWNERS * MaxLogChunksPerOwnerMultiplier / MaxLogChunksPerOwnerDivisor) log chunks
    // We call it cutThreshold
    optional uint64 MinLogChunksTotal = 1;
    optional uint64 MaxLogChunksPerOwnerMultiplier = 2;
    optional uint64 MaxLogChunksPerOwnerDivisor = 3;

    // Once SortFreeChunksPerItems chunks are released, free chuks are sorted
    optional uint64 SortFreeChunksPerItems = 4;


    optional uint32 SectorSize = 6;
    optional uint32 ChunkSize = 7;


    optional ESwitch GetDriveDataSwitch = 1000;  // Disable is same as DoNotTouch.
    optional ESwitch WriteCacheSwitch = 1001;  // non-force versions need GetDriveDataSwitch enabled to work.


    optional uint64 DriveModelSeekTimeNs = 1005;
    optional uint64 DriveModelSpeedBps = 1006;
    optional uint64 DriveModelBulkWrieBlockSize = 1008;
    optional uint64 DriveModelTrimSpeedBps = 1009;
    optional uint64 ReorderingMs = 1011;
    optional uint64 CostLimitNs = 1022;
    optional uint64 DeviceInFlight = 1023;

    optional uint32 BufferPoolBufferSizeBytes = 1024;
    optional uint32 BufferPoolBufferCount = 1025;
    optional uint32 MaxQueuedCompletionActions = 1026;

    optional uint64 InsaneLogChunksMultiplier = 2000;  // Log of cutThreshold * InsaneLogChunksMultiplier is insane

    optional uint64 ExpectedSlotCount = 2001; // Number of slots to calculate per-vdisk disk space limit.

    optional uint32 ChunkBaseLimit = 2002; // Free chunk permille that triggers Cyan color (e.g. 100 is 10%). Between 130 (default) and 13.

    optional uint32 CompletionThreadsCount = 2003;
};
